### PR TITLE
MergeTree read split ranges into intersecting and non intersecting injection

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added MergeTree read split ranges into intersecting and non intersecting fault injection using `merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_fault_probability` setting.

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 0da817aae1768b884cd282c77e424cd20e3742f1 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60548/0da817aae1768b884cd282c77e424cd20e3742f1/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Core/Settings.h [hunk 1]
```diff
@@ -284,6 +284,7 @@ class IColumn;
     M(UInt64, read_backoff_min_concurrency, 1, "Settings to try keeping the minimal number of threads in case of slow reads.", 0) \
     \
     M(Float, memory_tracker_fault_probability, 0., "For testing of `exception safety` - throw an exception every time you allocate memory with the specified probability.", 0) \
+    M(Float, merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_fault_probability, 1.0, "For testing of `PartsSplitter` - split read ranges into intersecting and non intersecting every time you read from MergeTree with the specified probability.", 0) \
```
Comment (azat) [inline src/Core/Settings.h:287]: Shouldn't this be default to `0` and enabled for tests via `tests/config/config.d/`?

Comment (kitaisreal) [reply src/Core/Settings.h:287]: We first need to set it to 1.0 and then fix all tests that are broken.
After this we will enable it only for tests via randomized settings.

File: src/Processors/QueryPlan/PartsSplitter.cpp [hunk 2]
```diff
@@ -30,6 +34,71 @@ std::string toString(const Values & value)
...
+        default:
+        {
+            break;
+        }
+    }
+
+    return true;
+}
+
+bool isSafePrimaryKey(const KeyDescription & primary_key)
```
Comment (nickitat) [inline src/Processors/QueryPlan/PartsSplitter.cpp:91]: maybe clarify in a comment, for what it should be safe

File: src/Processors/QueryPlan/PartsSplitter.cpp [hunk 3]
```diff
@@ -216,18 +284,82 @@ struct PartsRangesIterator
...
+    {}
+
+    bool operator==(const PartRangeIndex & other) const
+    {
+        return part_index == other.part_index && range.begin == other.range.begin && range.end == other.range.end;
+    }
+
+    bool operator<(const PartRangeIndex & other) const
+    {
+        return part_index < other.part_index && range.begin < other.range.begin && range.end < other.range.end;
```
Comment (nickitat) [inline src/Processors/QueryPlan/PartsSplitter.cpp:323]: why it is not equivalent to `tuple(part_index, begin, end) < tuple(other...)`? pls leave a comment.

**Review by nickitat — COMMENTED** [review]
* IMO, it would be enough to turn this on only for a few selected tests. but if you already fixed everything - it's cool to have max possible "blast radius"
* I think it is a bit misleading to call this feature _fault_ injection. 'cause we are not aiming to produce any faults. maybe remove "fault" and leave only "injection"?

**Comment by kitaisreal** [comment]
> * IMO, it would be enough to turn this on only for a few selected tests. but if you already fixed everything - it's cool to have max possible "blast radius"
> * I think it is a bit misleading to call this feature _fault_ injection. 'cause we are not aiming to produce any faults. maybe remove "fault" and leave only "injection"?

I think we need to enable it for all tests. I already found some issues with NaNs and Nulls.
I agree let's rename setting into  `merge_tree_read_split_ranges_into_intersecting_and_non_intersecting_injection_probability`.

File: src/Processors/QueryPlan/PartsSplitter.cpp (hunk 2 above)
Comment (kitaisreal) [reply src/Processors/QueryPlan/PartsSplitter.cpp]: Updated.

File: src/Processors/QueryPlan/PartsSplitter.cpp (hunk 3 above)
Comment (kitaisreal) [reply src/Processors/QueryPlan/PartsSplitter.cpp]: Updated.

**Review by nickitat — APPROVED** [review]

**Comment by kitaisreal** [comment]
Setup:
```
CREATE TABLE test_table_1 (id UInt64, value String) ENGINE=MergeTree ORDER BY id;
CREATE TABLE test_table_2 (id UInt64, value String) ENGINE=MergeTree ORDER BY id;

INSERT INTO test_table_1 SELECT number, number FROM numbers(150000000);
INSERT INTO test_table_2 SELECT number, number FROM numbers(150000000);
```

INNER JOIN:
```
SELECT
    lhs.id,
    rhs.id,
    lhs.value,
    rhs.value
FROM test_table_1 AS lhs
INNER JOIN test_table_2 AS rhs ON lhs.id = rhs.id
WHERE lhs.id = 5
FORMAT `Null`
```

Before:
```
0 rows in set. Elapsed: 22.130 sec. Processed 150.01 million rows, 3.79 GB (6.78 million rows/s., 171.21 MB/s.)
Peak memory usage: 18.41 GiB.
```

After:
```
0 rows in set. Elapsed: 0.005 sec. Processed 16.38 thousand rows, 131.19 KB (3.21 million rows/s., 25.73 MB/s.)
Peak memory usage: 579.28 KiB.

```

LEFT JOIN:
```
SELECT
    lhs.id,
    rhs.id,
    lhs.value,
    rhs.value
FROM test_table_1 AS lhs
LEFT JOIN test_table_2 AS rhs ON lhs.id = rhs.id
WHERE lhs.id = 5
FORMAT `Null`
```

Before:
```
0 rows in set. Elapsed: 22.680 sec. Processed 150.01 million rows, 3.79 GB (6.61 million rows/s., 167.06 MB/s.)
Peak memory usage: 18.42 GiB.
```

After:
```
0 rows in set. Elapsed: 0.005 sec. Processed 16.38 thousand rows, 131.19 KB (3.30 million rows/s., 26.45 MB/s.)
Peak memory usage: 579.28 KiB.
```

Right JOIN:
```
SELECT
    lhs.id,
    rhs.id,
    lhs.value,
    rhs.value
FROM test_table_1 AS lhs
RIGHT JOIN test_table_2 AS rhs ON lhs.id = rhs.id
WHERE lhs.id = 5
FORMAT `Null`
```

Before:
```
0 rows in set. Elapsed: 29.867 sec. Processed 300.00 million rows, 7.58 GB (10.04 million rows/s., 253.72 MB/s.)
Peak memory usage: 20.34 GiB.
```

After:
```
0 rows in set. Elapsed: 0.005 sec. Processed 16.38 thousand rows, 131.19 KB (3.13 million rows/s., 25.06 MB/s.)
Peak memory usage: 585.64 KiB.
```

FULL OUTER JOIN:
```
SELECT
    lhs.id,
    rhs.id,
    lhs.value,
    rhs.value
FROM test_table_1 AS lhs
FULL OUTER JOIN test_table_2 AS rhs ON lhs.id = rhs.id
WHERE lhs.id = 5
FORMAT `Null`
```

Before:
```
0 rows in set. Elapsed: 32.153 sec. Processed 300.00 million rows, 7.58 GB (9.33 million rows/s., 235.68 MB/s.)
Peak memory usage: 20.35 GiB.
```

After:
```
0 rows in set. Elapsed: 0.005 sec. Processed 16.38 thousand rows, 131.19 KB (3.44 million rows/s., 27.56 MB/s.)
Peak memory usage: 585.64 KiB.
```

**Review by nickitat — APPROVED** [review]

**Pull request merged**